cmd/bucket-policy-handlers.go

```diff
@@ -192,13 +192,38 @@ func (api objectAPIHandlers) PutBucketPolicyHandler(w http.ResponseWriter, r *ht
 		return
 	}
 
-	// Set the bucket policy in memory.
-	globalBucketPolicies.SetBucketPolicy(bucket, policy)
-
 	// Success.
 	writeSuccessNoContent(w)
 }
 
+// PutBucketPolicy - Persist the given policy to object layer, and
+// notify nodes in the cluster about the change. In-memory state is
+// updated in response to the notification. Assumes that the given
+// bucketPolicy is validated or is nil. If it is nil, bucket policy is
+// removed.
+func PutBucketPolicy(bucket string, bktpolicy *bucketPolicy, objAPI ObjectLayer) error {
+	// FIXME: Race exists between the bucket existence check and
+	// then updating the bucket policy.
+	if err := isBucketExist(bucket, objAPI); err != nil {
+		return err
+	}
+
+	// persist to object layer
+	if bktpolicy == nil {
```

Comment:
Ok I've stopped using `nil` as a zero value here and added a `.IsEmpty` for the struct to identify an empty policy. This lets the behaviour be uniform even in case an empty policy was sent to PutBucketPolicy API. Additionally, we repeat less code by not introducing a high level remove policy function (there is already one at the lower level to delete the policy from storage).